blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,449 @@
+    Title: Racket v6.11
+    Date: 2017-11-XXTXX:XX:XX
+    Tags: draft
+
+*posted by Andrew Kent*
+
+# Refinement and Dependent Function Types in Typed Racket
+
+As of Racket v6.11, Typed Racket now supports basic refinement and
+dependent function types. This post gives an overview for working with
+these types while writing some simple vector operations.
+
+At the time of writing this post, these types are documented under
+Typed Racket's [experimental
+features](https://docs.racket-lang.org/ts-reference/Experimental_Features.html). They
+are experimental in the sense that they are relatively new. We look
+forward to improving their usefulness by incorporating user feedback.
+
+In the near future, if you wish to experiment with refinement and/or
+dependent function types you may want to grab a [snapshot
+build](https://pre.racket-lang.org/installers/), as it will have
+additional features and/or bug fixes from user feedback.
+
+
+## Refinement Types
+
+Let us examine the type of `vector-ref` in Typed Racket:
+
+```racket
+(All (a) (-> (Vectorof a) Integer a))
+```
+
+It states that the index (the second argument) must be an `Integer`,
+i.e. any whole-number value in the exclusive range (-∞ , ∞). This
+description, however, is somewhat coarse: we obviously cannot use
+_any_ integer. To avoid a runtime error when indexing into a vector
+`v`, we must use an integer in the range [0, `(vector-length
+v)`). With a refinement type we can precisely describe this subset of
+the integers:
+
+```racket
+(Refine [i : Integer] (<= 0 i (- (vector-length v) 1)))
+```
+
+This refinement type describes values `i` of type `Integer` such that
+`(<= 0 i (- (vector-length v) 1))` produces `#true`. The documentation
+for [Refine](http://docs.racket-lang.org/search/index.html?q=Refine)
```

Comment:
Do you really want to link to the search page?

Link for `Refine` in typed racket:
http://docs.racket-lang.org/ts-reference/Experimental_Features.html#%28form._%28%28lib._typed-racket%2Fbase-env%2Fbase-types-extra..rkt%29._.Refine%29%29

Reply:
If Refine moves from the Experimental Features category to somewhere else that link would break -- linking to the search seemed a simple way to "future proof" it against that change.